src/parse.rs

```diff
@@ -0,0 +1,2602 @@
+) -> Result<Option<A>, ParseError> {
+    let mut next_op = next_op;
+    loop {
+        match (opstack.pop(), next_op.take()) {
+            (None, None) => {
+                if let Some(fin) = estack.pop() {
+                    if estack.is_empty() {
+                        return Ok(Some(fin));
+                    } else {
+                        panic!(parser_panic_text("Expression not fully reduced."))
+                    }
+                } else {
+                    return Ok(None);
+                }
+            }
+
+            (None, Some(op)) => {
+                opstack.push(op);
+                break;
+            }
+
+            (Some((op, _)), None) => do_reduce(op, estack),
+
+            (Some((opl, pl)), Some((opr, pr))) => {
+                match pl.cmp(&pr) {
+                    // all ops are left associative
+                    Ordering::Greater | Ordering::Equal => {
+                        do_reduce(opl, estack);
+                        next_op = Some((opr, pr));
+                    }
+                    Ordering::Less => {
+                        opstack.push((opl, pl));
+                        opstack.push((opr, pr));
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    Ok(None)
+}
+
+fn precedence(t: &Tok) -> Option<u8> {
+    if t == &Tok::Pipe {
+        return Some(5);
+    };
+    match tok_to_binop(t) {
+        Some(b) => Some(b.precedence()),
+        None => None,
+    }
+}
+
+fn tok_to_binop(t: &Tok) -> Option<BinOp> {
+    match *t {
+        Tok::VbarVbar => Some(BinOp::Or),
+        Tok::AmperAmper => Some(BinOp::And),
+        Tok::EqualEqual => Some(BinOp::Eq),
+        Tok::NotEqual => Some(BinOp::NotEq),
+        Tok::Less => Some(BinOp::LtInt),
+        Tok::LessEqual => Some(BinOp::LtEqInt),
+        Tok::Greater => Some(BinOp::GtInt),
+        Tok::GreaterEqual => Some(BinOp::GtEqInt),
+        Tok::LessDot => Some(BinOp::LtFloat),
+        Tok::LessEqualDot => Some(BinOp::LtEqFloat),
+        Tok::GreaterDot => Some(BinOp::GtFloat),
+        Tok::GreaterEqualDot => Some(BinOp::GtEqFloat),
+        Tok::Plus => Some(BinOp::AddInt),
+        Tok::Minus => Some(BinOp::SubInt),
+        Tok::PlusDot => Some(BinOp::AddFloat),
+        Tok::MinusDot => Some(BinOp::SubFloat),
+        Tok::Percent => Some(BinOp::ModuloInt),
+        Tok::Star => Some(BinOp::MultInt),
+        Tok::StarDot => Some(BinOp::MultFloat),
+        Tok::Slash => Some(BinOp::DivInt),
+        Tok::SlashDot => Some(BinOp::DivFloat),
+        _ => None,
+    }
+}
+// Simple-Op-Parser, perform reduction for expression
+fn do_reduce_expression(op: Spanned, estack: &mut Vec<UntypedExpr>) {
+    match (estack.pop(), estack.pop()) {
+        (Some(er), Some(el)) => {
+            let new_e = expr_op_reduction(op, el, er);
+            estack.push(new_e);
+        }
+        _ => panic!(parser_panic_text("Tried to reduce withot 2 expressions")),
```

Comment:
Can we swap these panics for `gleam::error::fatal_compiler_bug`? It prints a nice error message to the user.